pulsar/internal/connection.go

```diff
@@ -744,6 +745,46 @@ func (c *connection) handleAuthChallenge(authChallenge *pb.CommandAuthChallenge)
 	c.writeCommand(baseCommand(pb.BaseCommand_AUTH_RESPONSE, cmdAuthResponse))
 }
 
+func (c *connection) handleSendError(sendError *pb.CommandSendError, cmdError *pb.CommandError) {
+	c.log.Warnf("Received send error from server: [%v] : [%s]", sendError.GetError(), sendError.GetMessage())
+
+	requestID := cmdError.GetRequestId()
+	producerID := sendError.GetProducerId()
+
+	switch *sendError.Error {
+	case pb.ServerError_NotAllowedError:
+		c.pendingLock.Lock()
```

Comment:
This block of code is used in a couple of places now maybe we should add a function?  Something like this?

```
func deletePendingRequest(requestID) (request, bool) {
   c.pendingLock.Lock()
   defer c.pendingLock.Unlock()
   request, ok := c.pendingReqs[requestID]
    if ok {
	delete(c.pendingReqs, requestID)
   }
   return request, ok
		
}
```

Reply:
Good ideas for this.

Reply:
> The java client closes the connection in the default case.
> 
> https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java#L649
> 
> Should we do the same. What should the expected client behavior be here?

About this, Look at the comments like reconnect, look at the code like close the current connection:

```
            // By default, for transient error, let the **reconnection** logic
            // to take place and re-establish the produce again
            ctx.close();
```